pkg/client/block/client.go

```diff
@@ -46,32 +50,142 @@ func NewBlockClient(
 	if err != nil {
 		return nil, err
 	}
-	return &blockClient{eventsReplayClient: client}, nil
+
+	// latestBlockPublishCh is the channel that notifies the latestBlockReplayObs of a
+	// new block, whether it comes from a direct query or an event subscription query.
+	latestBlockReplayObs, latestBlockPublishCh := channel.NewReplayObservable[client.Block](ctx, defaultBlocksReplayLimit)
+
+	blockClient := &blockReplayClient{
+		eventsReplayClient:   eventsReplayClient,
+		latestBlockReplayObs: latestBlockReplayObs,
+		latestBlockPublishCh: latestBlockPublishCh,
+	}
+
+	if err := depinject.Inject(deps, &blockClient.queryClient); err != nil {
+		return nil, err
+	}
+
+	blockClient.forEachBlockEvent(ctx)
+
+	if err := blockClient.getInitialBlock(ctx); err != nil {
+		return nil, err
+	}
+
+	return blockClient, nil
 }
 
-// blockClient is a wrapper around an EventsReplayClient that implements the
+// blockReplayClient is a wrapper around an EventsReplayClient that implements the
 // BlockClient interface for use with cosmos-sdk networks.
-type blockClient struct {
+type blockReplayClient struct {
+	// queryClient is the RPC client that is used to query for the initial block
+	// upon blockClient construction. The result of this query is only used if it
+	// returns before the eventsReplayClient receives its first event.
+	queryClient cometclient.CometRPC
```

Comment:
Should we rename this to `oneTimeUseQueryClient` or `onStartQueryClient` to be ultra explicit?